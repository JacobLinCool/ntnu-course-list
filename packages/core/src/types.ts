export type NumberLike = number | `${number}`;
export type YesNo = "Y" | "N";

export interface CourseListFetcherOptions {
	year: NumberLike;
	term: NumberLike;
	deptCode1?: string;
	zuDeptCode1?: string;
	kind?: string;
	generalCore?: string;
	chn?: string;
	teacher?: string;
	engTeach?: YesNo;
	moocs_v?: YesNo;
	remoteCourse?: YesNo;
	digital?: YesNo;
	adsl?: YesNo;
	classCode?: string;
	language1?: "chinese" | "english";
	serial_number?: NumberLike;
	course_code?: string;
	download?: "xls";
}

export interface CourseJSON {
	開課序號: `${number}`;
	開課代碼: string;
	系所: string;
	組: string;
	年: string;
	班: string;
	全英語: string;
	MOOCS: string;
	限性別: string;
	中文課程名稱: string;
	英文課程名稱: string;
	學分: `${number}.${number}`;
	"必/選": "選" | "必";
	"全/半": "半" | "全";
	教師: string;
	地點時間: string;
	限修人數: `${number}`;
	選修人數: `${number}`;
	限修條件: string;
	備註: string;
}

export enum Day {
	一 = 1,
	二 = 2,
	三 = 3,
	四 = 4,
	五 = 5,
	六 = 6,
	日 = 7,
	None = 0,
}

export type Period = number | "A" | "B" | "C" | "D" | "E" | "F" | "G";

export interface ClassSchedule {
	day: Day;
	period: [start: Period, end: Period];
	location: string;
	classroom: string;
}
