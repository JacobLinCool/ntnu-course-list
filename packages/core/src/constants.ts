export const EXPORT_ENDPOINT =
	"https://courseap2.itc.ntnu.edu.tw/acadmOpenCourse/CofopdlByAcadmRpt";
