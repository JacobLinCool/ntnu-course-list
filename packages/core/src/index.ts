export * from "./constants";
export * from "./fetcher";
export * from "./parse";
export * from "./pkg";
export * from "./types";
