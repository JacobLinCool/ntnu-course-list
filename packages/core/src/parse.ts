import type { ClassSchedule, Period } from "./types";
import { Day } from "./types";

const NORMAL_FORMAT =
	/(\p{Script=Han})\s+([\dABCDEFG]+(-[\dABCDEFG]+)?)\s+((\p{Script=Han}+?)\s+)?([^\s,]+)/gu;
const TIME_ONLY_FORMAT_1 = /(\p{Script=Han}),(\d+)-\p{Script=Han},(\d+)/gu;
const TIME_ONLY_FORMAT_2 = /(\p{Script=Han})\s+(\d+(-\d+)?)/gu;

/**
 * Parse the normal schedule format.
 */
function parse_1(raw: string): ClassSchedule[] {
	const result: ClassSchedule[] = [];
	let match: RegExpExecArray | null;

	while ((match = NORMAL_FORMAT.exec(raw)) !== null) {
		if (match.length !== 7) {
			continue;
		}

		const day = Day[match[1] as keyof typeof Day];
		const period_raw = match[2].split("-");
		const start = period_raw[0],
			end = period_raw[1] || period_raw[0];
		const period: [Period, Period] = [
			/[ABCDEFG]/.test(start) ? (start as Period) : parseInt(start),
			/[ABCDEFG]/.test(end) ? (end as Period) : parseInt(end),
		];
		const location = match[5] || "";
		const classroom = match[6];

		result.push({
			day,
			period,
			location,
			classroom,
		});
	}

	return result;
}

/**
 * Parse the time only format 1.
 */
function parse_2(raw: string): ClassSchedule[] {
	const result: ClassSchedule[] = [];
	let match: RegExpExecArray | null;

	while ((match = TIME_ONLY_FORMAT_1.exec(raw)) !== null) {
		const day = Day[match[1] as keyof typeof Day];
		const period: [Period, Period] = [
			/[ABCDEFG]/.test(match[2]) ? (match[2] as Period) : parseInt(match[2]),
			/[ABCDEFG]/.test(match[3]) ? (match[3] as Period) : parseInt(match[3]),
		];

		result.push({
			day,
			period,
			location: "",
			classroom: "",
		});
	}

	return result;
}

/**
 * Parse the time only format 2.
 */
function parse_3(raw: string): ClassSchedule[] {
	const result: ClassSchedule[] = [];
	let match: RegExpExecArray | null;

	while ((match = TIME_ONLY_FORMAT_2.exec(raw)) !== null) {
		const day = Day[match[1] as keyof typeof Day];
		const period_raw = match[2].split("-");
		const start = period_raw[0],
			end = period_raw[1] || period_raw[0];
		const period: [Period, Period] = [
			/[ABCDEFG]/.test(start) ? (start as Period) : parseInt(start),
			/[ABCDEFG]/.test(end) ? (end as Period) : parseInt(end),
		];

		result.push({
			day,
			period,
			location: "",
			classroom: "",
		});
	}

	return result;
}

/**
 * Parse the single string format.
 */
function parse_4(raw: string): ClassSchedule[] {
	raw = raw.trim();
	if (raw && !raw.includes(" ")) {
		return [
			{
				day: Day.None,
				period: [0, 0],
				location: "",
				classroom: raw,
			},
		];
	}
	return [];
}

/**
 * Parse the schedule string into an array of {@link ClassSchedule}.
 * @param raw The schedule string. e.g. "三 7-9 公館 Ｂ101" or "二 2 公館 Ｅ101, 五 8-9 公館 Ｅ101"
 */
export function parse_schedule(raw: string): ClassSchedule[] {
	let result = parse_1(raw);
	if (result.length === 0) {
		result = parse_2(raw);
	}
	if (result.length === 0) {
		result = parse_3(raw);
	}
	if (result.length === 0) {
		result = parse_4(raw);
	}

	return result;
}
