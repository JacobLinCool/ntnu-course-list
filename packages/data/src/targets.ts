import type { CourseListFetcherOptions } from "ntnu-course-list";

export const targets: CourseListFetcherOptions[] = [
	{
		year: 110,
		term: 1,
	},
	{
		year: 110,
		term: 2,
	},
	{
		year: 110,
		term: 3,
	},
	{
		year: 111,
		term: 1,
	},
	{
		year: 111,
		term: 2,
	},
	{
		year: 111,
		term: 3,
	},
	{
		year: 112,
		term: 1,
	},
	{
		year: 112,
		term: 2,
	},
	{
		year: 112,
		term: 3,
	},
	{
		year: 113,
		term: 1,
	},
	{
		year: 113,
		term: 2,
	},
	{
		year: 113,
		term: 3,
	},
];

export function filter() {
	const filter = process.argv[2];
	if (filter) {
		const [year, term] = filter.split("-").map((x) => parseInt(x));
		if (year && term) {
			for (let i = targets.length - 1; i >= 0; i--) {
				if (targets[i].year !== year || targets[i].term !== term) {
					targets.splice(i, 1);
				}
			}
		} else if (year) {
			for (let i = targets.length - 1; i >= 0; i--) {
				if (targets[i].year !== year) {
					targets.splice(i, 1);
				}
			}
		} else {
			throw new Error("Invalid filter, use format like 112-1 or 112");
		}
	}
}
