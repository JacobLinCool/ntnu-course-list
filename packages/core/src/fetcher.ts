import fetch from "cross-fetch";
import debug from "debug";
import { read, utils } from "xlsx";
import { EXPORT_ENDPOINT } from "./constants";
import type { CourseJSON, CourseListFetcherOptions } from "./types";

let count = 0;

export class CourseListFetcher {
	protected log = debug(`ntnu-course-list:fetcher:${count++}`);
	protected endpoint: string;

	constructor(endpoint = EXPORT_ENDPOINT) {
		this.endpoint = endpoint;
		this.log("Initialized with endpoint: %s", endpoint);
	}

	/**
	 * Get course list in xls format.
	 */
	public async xls(opt: CourseListFetcherOptions): Promise<ArrayBuffer> {
		const body = `rpt=cofopdl&acadmYear=${opt.year}&acadmTerm=${opt.term}&deptCode1=${
			opt.deptCode1 ?? ""
		}&zuDeptCode1=${opt.zuDeptCode1 ?? ""}&kind=${opt.kind ?? ""}&generalCore=${
			opt.generalCore ?? ""
		}&chn=&teacher=${opt.teacher ?? ""}&engTeach=${opt.engTeach ?? "N"}&moocs_v=${
			opt.moocs_v ?? "N"
		}&remoteCourse=${opt.remoteCourse ?? "N"}&digital=${opt.digital ?? "N"}&adsl=${
			opt.adsl ?? "N"
		}&classCode=${opt.classCode ?? ""}&language1=${opt.language1 ?? "chinese"}&serial_number=${
			opt.serial_number ?? ""
		}&course_code=${opt.course_code ?? ""}&download=${opt.download ?? "xls"}`;
		this.log("Fetching course list with body: %s", body);

		const res = await fetch(this.endpoint, {
			method: "POST",
			headers: {
				"Content-Type": "application/x-www-form-urlencoded",
			},
			body,
		});
		if (res.headers.get("content-type")?.includes("text/html")) {
			const text = await res.text();
			this.log(text);
			throw new Error(`Failed to fetch course list: ${text}`);
		}

		const buffer = await res.arrayBuffer();
		return buffer;
	}

	/**
	 * Get course list in json format.
	 */
	public async json(opt: CourseListFetcherOptions): Promise<CourseJSON[]> {
		const xls = await this.xls(opt);
		const workbook = read(xls, { type: "array" });
		const sheet = workbook.Sheets[workbook.SheetNames[0]];
		const json = utils.sheet_to_json<CourseJSON>(sheet, { raw: true });
		return json;
	}

	/**
	 * Toggle debug mode.
	 * @param enabled Whether to enable debug mode.
	 */
	public debug(enabled = true) {
		if (enabled) {
			this.log.enabled = true;
		} else {
			this.log.enabled = false;
		}
	}
}
