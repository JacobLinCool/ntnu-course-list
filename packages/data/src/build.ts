import fs from "node:fs";
import path from "node:path";
import { CourseJSON, CourseListFetcher, parse_schedule } from "ntnu-course-list";
import { Parsed } from "../types";
import { filter, targets } from "./targets";

main();

async function main() {
	filter();
	console.log({ targets });

	const tag = "Build Package";
	console.time(tag);

	await prepare();
	console.timeLog(tag, "Prepared");

	await download();
	console.timeLog(tag, "Downloaded");

	await parse();
	console.timeLog(tag, "Parsed");

	await split();
	console.timeLog(tag, "Splitted");

	await pack();
	console.timeLog(tag, "Packaged");
}

async function prepare() {
	const dir = path.join(__dirname, "../_package");
	for (const target of targets) {
		const root = path.join(dir, `${target.year}${target.term}`);
		if (!fs.existsSync(root)) {
			fs.mkdirSync(root, { recursive: true });
		}
	}
}

async function download() {
	for (const target of targets) {
		const file = path.join(__dirname, `../_package/${target.year}${target.term}/raw.json`);
		if (fs.existsSync(file)) {
			console.log(`Skipping ${target.year} ${target.term}`);
			continue;
		}

		const fetcher = new CourseListFetcher();
		fetcher.debug(true);

		const json = await fetcher.json(target);
		fs.writeFileSync(file, JSON.stringify(json));
	}
}

async function parse() {
	for (const target of targets) {
		const json = JSON.parse(
			fs.readFileSync(
				path.join(__dirname, `../_package/${target.year}${target.term}/raw.json`),
				"utf-8",
			),
		) as CourseJSON[];
		let good = 0;
		const parsed = json.map((course) => {
			const parsed = parse_schedule(course["地點時間"]);
			if (parsed.length > 0 || course["地點時間"] === "") {
				good++;
			}
			return { ...course, 地點時間: parsed };
		});
		fs.writeFileSync(
			path.join(__dirname, `../_package/${target.year}${target.term}/all.json`),
			JSON.stringify(parsed),
		);

		console.log(
			`Parsed ${target.year} ${target.term}: ${good} / ${parsed.length} (${(
				(good / parsed.length) *
				100
			).toFixed(2)}%)`,
		);
	}
}

async function split() {
	// split by department
	for (const target of targets) {
		const json = JSON.parse(
			fs.readFileSync(
				path.join(__dirname, `../_package/${target.year}${target.term}/all.json`),
				"utf-8",
			),
		) as Parsed[];

		const depts = new Set(json.map((course) => course.系所));
		for (const dept of depts) {
			const courses = json.filter((course) => course.系所 === dept);
			const dir = path.join(__dirname, `../_package/${target.year}${target.term}/${dept}`);
			if (!fs.existsSync(dir)) {
				fs.mkdirSync(dir, { recursive: true });
			}
			fs.writeFileSync(path.join(dir, "all.json"), JSON.stringify(courses));
		}

		fs.writeFileSync(
			path.join(__dirname, `../_package/${target.year}${target.term}/depts.json`),
			JSON.stringify([...depts]),
		);
	}
}

async function pack() {
	const base_pkg = JSON.parse(fs.readFileSync(path.join(__dirname, "../package.json"), "utf-8"));
	const base_readme = fs.readFileSync(path.join(__dirname, "../README.md"), "utf-8");
	const types = fs.readFileSync(path.join(__dirname, "../types.d.ts"), "utf-8");

	for (const target of targets) {
		const dir = path.join(__dirname, `../_package/${target.year}${target.term}`);

		// build README.md
		fs.writeFileSync(
			path.join(dir, "README.md"),
			`# NTNU Course List ${target.year} 學年度第 ${target.term} 學期\n\n---\n\n${base_readme}`,
		);

		// build package.json
		const major = 1;
		const revision = (await get_last_revision(major, `${target.year}${target.term}`)) + 1;
		delete base_pkg.scripts;
		delete base_pkg.devDependencies;
		base_pkg.version = `${major}.${target.year}${target.term}.${revision}`;
		base_pkg.description = `NTNU Course List ${target.year} 學年度第 ${target.term} 學期`;
		base_pkg.private = false;
		fs.writeFileSync(path.join(dir, "package.json"), JSON.stringify(base_pkg, null, 4));

		// copy types.d.ts
		fs.writeFileSync(path.join(dir, "types.d.ts"), types);
	}
}

async function get_last_revision(major: number, minor: string) {
	const res = await fetch(`https://registry.npmjs.org/ntnu-course-list-data/${major}.${minor}.*`);
	if (!res.ok) {
		return -1;
	}
	const json = await res.json();
	const revision = parseInt(json.version.split(".")[2]);
	return revision;
}
