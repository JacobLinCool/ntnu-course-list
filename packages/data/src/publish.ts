import { execSync } from "node:child_process";
import fs from "node:fs";
import path from "node:path";
import { filter, targets } from "./targets";

main();

async function main() {
	filter();
	console.log({ targets });

	for (const target of targets) {
		const dir = path.join(__dirname, "../_package", `${target.year}${target.term}`);
		if (!fs.existsSync(dir)) {
			throw new Error(`Directory ${dir} does not exist`);
		}
		if (!fs.existsSync(path.join(dir, "package.json"))) {
			throw new Error(`File ${path.join(dir, "package.json")} does not exist`);
		}
		const pkg = JSON.parse(fs.readFileSync(path.join(dir, "package.json"), "utf-8"));
		console.log(`Publishing ${pkg.name}@${pkg.version}...`);

		execSync("npm publish", { cwd: dir, stdio: "inherit" });
	}
}
