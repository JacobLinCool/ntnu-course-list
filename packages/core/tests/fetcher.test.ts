import { expect, it } from "vitest";
import { CourseListFetcher } from "../src/fetcher";

it("should fetch the course list in xls", async () => {
	const fetcher = new CourseListFetcher();
	const options = {
		year: 111,
		term: 2,
		deptCode1: "SU47",
	};

	const result = await fetcher.xls(options);

	expect(result).toBeInstanceOf(ArrayBuffer);
	expect(result.byteLength).toBeGreaterThan(0);
});

it("should fetch the course list in json", async () => {
	const fetcher = new CourseListFetcher();
	const options = {
		year: 111,
		term: 2,
		deptCode1: "SU47",
	};

	const result = await fetcher.json(options);

	expect(result).toMatchSnapshot();
});

it("should failed to fetch non-exist course list", async () => {
	const fetcher = new CourseListFetcher();
	const options = {
		year: 1000,
		term: 1,
		deptCode1: "SU47",
	};

	await expect(fetcher.xls(options)).rejects.toThrowError();
});
