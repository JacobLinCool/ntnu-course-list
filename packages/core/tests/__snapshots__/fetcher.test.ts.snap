// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should fetch the course list in json 1`] = `
[
  {
    "MOOCS": "",
    "中文課程名稱": "資訊科技教學實習（教）",
    "備註": "1.需先修習通過資訊科技教材教法(教)課程。 2.請務必依【教育專門課程對應科目】修習課程，避免造成學分無法認定之情形發生。",
    "全/半": "半",
    "全英語": "",
    "地點時間": "五 6-9 公館 Ｓ506",
    "學分": "2.0",
    "年": "",
    "必/選": "選",
    "教師": "柯佳伶 方瓊瑤 紀博文",
    "班": "",
    "系所": "資工系（學）",
    "組": "",
    "英文課程名稱": "Teaching Practicum: Computer Science and Information Technology",
    "選修人數": "15",
    "開課代碼": "CSUE310",
    "開課序號": "2873",
    "限修人數": "33",
    "限修條件": "",
    "限性別": "",
  },
  {
    "MOOCS": "",
    "中文課程名稱": "物件導向分析與設計",
    "備註": "",
    "全/半": "半",
    "全英語": "",
    "地點時間": "五 2-4 公館 Ｂ102",
    "學分": "3.0",
    "年": "",
    "必/選": "選",
    "教師": "林均翰",
    "班": "大碩",
    "系所": "資工系（學）",
    "組": "",
    "英文課程名稱": "Object-Oriented Analysis and Design",
    "選修人數": "49",
    "開課代碼": "CSC0005",
    "開課序號": "2874",
    "限修人數": "50",
    "限修條件": "",
    "限性別": "",
  },
  {
    "MOOCS": "",
    "中文課程名稱": "無線通訊",
    "備註": "",
    "全/半": "半",
    "全英語": "是",
    "地點時間": "二 7-9 公館 Ｂ102",
    "學分": "3.0",
    "年": "",
    "必/選": "選",
    "教師": "賀耀華",
    "班": "大碩",
    "系所": "資工系（學）",
    "組": "",
    "英文課程名稱": "Wireless Communications",
    "選修人數": "37",
    "開課代碼": "CSC0010",
    "開課序號": "2875",
    "限修人數": "50",
    "限修條件": "",
    "限性別": "",
  },
  {
    "MOOCS": "",
    "中文課程名稱": "電腦輔助VLSI設計",
    "備註": "",
    "全/半": "半",
    "全英語": "",
    "地點時間": "五 7-9 公館 Ｂ101",
    "學分": "3.0",
    "年": "",
    "必/選": "選",
    "教師": "黃文吉",
    "班": "大碩",
    "系所": "資工系（學）",
    "組": "",
    "英文課程名稱": "Computer-Aided Design for VLSI Design",
    "選修人數": "59",
    "開課代碼": "CSC0011",
    "開課序號": "2876",
    "限修人數": "50",
    "限修條件": "",
    "限性別": "",
  },
  {
    "MOOCS": "",
    "中文課程名稱": "資訊安全",
    "備註": "",
    "全/半": "半",
    "全英語": "",
    "地點時間": "一 2-4 公館 Ｂ102",
    "學分": "3.0",
    "年": "",
    "必/選": "選",
    "教師": "紀博文",
    "班": "大碩",
    "系所": "資工系（學）",
    "組": "",
    "英文課程名稱": "Information Security",
    "選修人數": "68",
    "開課代碼": "CSC0072",
    "開課序號": "2877",
    "限修人數": "50",
    "限修條件": "",
    "限性別": "",
  },
  {
    "MOOCS": "",
    "中文課程名稱": "即時系統",
    "備註": "",
    "全/半": "半",
    "全英語": "是",
    "地點時間": "四 7-9 公館 Ｓ204",
    "學分": "3.0",
    "年": "",
    "必/選": "選",
    "教師": "王超",
    "班": "大碩",
    "系所": "資工系（學）",
    "組": "",
    "英文課程名稱": "Real-Time Systems",
    "選修人數": "18",
    "開課代碼": "CSC9006",
    "開課序號": "2878",
    "限修人數": "50",
    "限修條件": "",
    "限性別": "",
  },
  {
    "MOOCS": "",
    "中文課程名稱": "應用密碼學",
    "備註": "",
    "全/半": "半",
    "全英語": "",
    "地點時間": "四 2-4 公館 Ｓ204",
    "學分": "3.0",
    "年": "",
    "必/選": "選",
    "教師": "官振傑",
    "班": "大碩",
    "系所": "資工系（學）",
    "組": "",
    "英文課程名稱": "Applied Cryptography",
    "選修人數": "49",
    "開課代碼": "CSC9010",
    "開課序號": "2879",
    "限修人數": "50",
    "限修條件": "",
    "限性別": "",
  },
  {
    "MOOCS": "",
    "中文課程名稱": "程式設計（二）",
    "備註": "",
    "全/半": "半",
    "全英語": "",
    "地點時間": "一 7 公館 理圖807, 三 8-9 公館 理圖807",
    "學分": "3.0",
    "年": "1",
    "必/選": "必",
    "教師": "紀博文",
    "班": "",
    "系所": "資工系（學）",
    "組": "A",
    "英文課程名稱": "Computer Programming (II)",
    "選修人數": "70",
    "開課代碼": "CSU0002",
    "開課序號": "2880",
    "限修人數": "70",
    "限修條件": "◎必須先修過【CSU0001  程式設計（一）】",
    "限性別": "",
  },
  {
    "MOOCS": "",
    "中文課程名稱": "程式設計（二）",
    "備註": "",
    "全/半": "半",
    "全英語": "",
    "地點時間": "二 2-4 公館 理圖807",
    "學分": "3.0",
    "年": "1",
    "必/選": "必",
    "教師": "張鈞法",
    "班": "",
    "系所": "資工系（學）",
    "組": "B",
    "英文課程名稱": "Computer Programming (II)",
    "選修人數": "46",
    "開課代碼": "CSU0002",
    "開課序號": "2881",
    "限修人數": "60",
    "限修條件": "◎必須先修過【CSU0001  程式設計（一）】",
    "限性別": "",
  },
  {
    "MOOCS": "",
    "中文課程名稱": "程式設計（二）",
    "備註": "1.本班為視訊上課，需自備筆電。2.第一周請至理圖807向老師索取授權碼。",
    "全/半": "半",
    "全英語": "",
    "地點時間": "一 7 公館 Ｓ601, 三 8-9 公館 Ｓ601",
    "學分": "3.0",
    "年": "1",
    "必/選": "必",
    "教師": "紀博文",
    "班": "",
    "系所": "資工系（學）",
    "組": "C",
    "英文課程名稱": "Computer Programming (II)",
    "選修人數": "43",
    "開課代碼": "CSU0002",
    "開課序號": "4800",
    "限修人數": "0",
    "限修條件": "◎必須先修過【CSU0001  程式設計（一）】",
    "限性別": "",
  },
  {
    "MOOCS": "",
    "中文課程名稱": "數位邏輯",
    "備註": "",
    "全/半": "半",
    "全英語": "",
    "地點時間": "二 7-9 公館 Ｅ202",
    "學分": "3.0",
    "年": "1",
    "必/選": "選",
    "教師": "吳榮根",
    "班": "",
    "系所": "資工系（學）",
    "組": "",
    "英文課程名稱": "Digital Logic",
    "選修人數": "99",
    "開課代碼": "CSU0009",
    "開課序號": "2882",
    "限修人數": "100",
    "限修條件": "",
    "限性別": "",
  },
  {
    "MOOCS": "",
    "中文課程名稱": "離散數學",
    "備註": "",
    "全/半": "半",
    "全英語": "",
    "地點時間": "一 8 公館 Ｅ102, 五 3-4 公館 Ｅ102",
    "學分": "3.0",
    "年": "1",
    "必/選": "選",
    "教師": "王弘倫",
    "班": "",
    "系所": "資工系（學）",
    "組": "",
    "英文課程名稱": "Discrete Mathematics",
    "選修人數": "102",
    "開課代碼": "CSU0011",
    "開課序號": "2883",
    "限修人數": "50",
    "限修條件": "",
    "限性別": "",
  },
  {
    "MOOCS": "",
    "中文課程名稱": "演算法",
    "備註": "",
    "全/半": "半",
    "全英語": "",
    "地點時間": "二 3-4 公館 Ｂ102, 四 2 公館 Ｂ102",
    "學分": "3.0",
    "年": "2",
    "必/選": "必",
    "教師": "葉梅珍",
    "班": "",
    "系所": "資工系（學）",
    "組": "A",
    "英文課程名稱": "Computer Algorithms",
    "選修人數": "80",
    "開課代碼": "CSU0018",
    "開課序號": "2884",
    "限修人數": "75",
    "限修條件": "◎必須先修過【CSU0001  程式設計（一）】",
    "限性別": "",
  },
  {
    "MOOCS": "",
    "中文課程名稱": "演算法",
    "備註": "",
    "全/半": "半",
    "全英語": "",
    "地點時間": "一 7-9 公館 Ｂ103",
    "學分": "3.0",
    "年": "2",
    "必/選": "必",
    "教師": "官振傑",
    "班": "",
    "系所": "資工系（學）",
    "組": "B",
    "英文課程名稱": "Computer Algorithms",
    "選修人數": "33",
    "開課代碼": "CSU0018",
    "開課序號": "2885",
    "限修人數": "50",
    "限修條件": "◎必須先修過【CSU0001  程式設計（一）】",
    "限性別": "",
  },
  {
    "MOOCS": "",
    "中文課程名稱": "計算機網路",
    "備註": "",
    "全/半": "半",
    "全英語": "",
    "地點時間": "二 8-9 公館 Ｂ101, 四 7 公館 Ｂ101",
    "學分": "3.0",
    "年": "2",
    "必/選": "選",
    "教師": "蔡榮宗",
    "班": "",
    "系所": "資工系（學）",
    "組": "",
    "英文課程名稱": "Computer Networks",
    "選修人數": "78",
    "開課代碼": "CSU0019",
    "開課序號": "2886",
    "限修人數": "80",
    "限修條件": "",
    "限性別": "",
  },
  {
    "MOOCS": "",
    "中文課程名稱": "數理統計",
    "備註": "",
    "全/半": "半",
    "全英語": "是",
    "地點時間": "一 2-4 公館 Ｂ101",
    "學分": "3.0",
    "年": "2",
    "必/選": "選",
    "教師": "陳翔瀚",
    "班": "",
    "系所": "資工系（學）",
    "組": "",
    "英文課程名稱": "Mathematical Statistics",
    "選修人數": "53",
    "開課代碼": "CSU0026",
    "開課序號": "2887",
    "限修人數": "50",
    "限修條件": "",
    "限性別": "",
  },
  {
    "MOOCS": "",
    "中文課程名稱": "計算機結構",
    "備註": "",
    "全/半": "半",
    "全英語": "",
    "地點時間": "二 2 公館 Ｂ102, 四 3-4 公館 Ｂ102",
    "學分": "3.0",
    "年": "2",
    "必/選": "必",
    "教師": "林均翰",
    "班": "",
    "系所": "資工系（學）",
    "組": "",
    "英文課程名稱": "Computer Architectures",
    "選修人數": "90",
    "開課代碼": "CSU0029",
    "開課序號": "2888",
    "限修人數": "50",
    "限修條件": "",
    "限性別": "",
  },
  {
    "MOOCS": "",
    "中文課程名稱": "進階程式設計",
    "備註": "期望同學需學習完基礎程式能力後再來修習本課程。",
    "全/半": "半",
    "全英語": "",
    "地點時間": "五 7-9 公館 理圖807",
    "學分": "3.0",
    "年": "3",
    "必/選": "選",
    "教師": "蔣宗哲",
    "班": "",
    "系所": "資工系（學）",
    "組": "",
    "英文課程名稱": "The Practice of Programming",
    "選修人數": "31",
    "開課代碼": "CSU0030",
    "開課序號": "2889",
    "限修人數": "50",
    "限修條件": "",
    "限性別": "",
  },
  {
    "MOOCS": "",
    "中文課程名稱": "工程數學",
    "備註": "",
    "全/半": "半",
    "全英語": "是",
    "地點時間": "二 2-4 公館 Ｂ103",
    "學分": "3.0",
    "年": "3",
    "必/選": "選",
    "教師": "王超",
    "班": "",
    "系所": "資工系（學）",
    "組": "",
    "英文課程名稱": "Engineering Mathematics",
    "選修人數": "17",
    "開課代碼": "CSU0032",
    "開課序號": "2890",
    "限修人數": "60",
    "限修條件": "",
    "限性別": "",
  },
  {
    "MOOCS": "",
    "中文課程名稱": "作業系統",
    "備註": "",
    "全/半": "半",
    "全英語": "是",
    "地點時間": "一 8-9 公館 Ｂ101, 五 3 公館 Ｂ101",
    "學分": "3.0",
    "年": "3",
    "必/選": "選",
    "教師": "賀耀華",
    "班": "",
    "系所": "資工系（學）",
    "組": "",
    "英文課程名稱": "Operating Systems",
    "選修人數": "77",
    "開課代碼": "CSU0033",
    "開課序號": "2891",
    "限修人數": "50",
    "限修條件": "",
    "限性別": "",
  },
  {
    "MOOCS": "",
    "中文課程名稱": "資料庫理論",
    "備註": "",
    "全/半": "半",
    "全英語": "",
    "地點時間": "一 3-4 公館 Ｅ101, 五 2 公館 理圖807",
    "學分": "3.0",
    "年": "3",
    "必/選": "選",
    "教師": "柯佳伶",
    "班": "",
    "系所": "資工系（學）",
    "組": "",
    "英文課程名稱": "Database Theories",
    "選修人數": "75",
    "開課代碼": "CSU0034",
    "開課序號": "2892",
    "限修人數": "50",
    "限修條件": "",
    "限性別": "",
  },
  {
    "MOOCS": "",
    "中文課程名稱": "資訊專題研究（二）：資訊理論",
    "備註": "",
    "全/半": "半",
    "全英語": "",
    "地點時間": "四 10-B 公館 教室自排",
    "學分": "3.0",
    "年": "3",
    "必/選": "選",
    "教師": "張鈞法 王超 陳俊祥",
    "班": "",
    "系所": "資工系（學）",
    "組": "",
    "英文課程名稱": "Special Topics on Computer Science (II) : Information Theories",
    "選修人數": "23",
    "開課代碼": "CSU0039",
    "開課序號": "2893",
    "限修人數": "50",
    "限修條件": "",
    "限性別": "",
  },
  {
    "MOOCS": "",
    "中文課程名稱": "資訊專題研究（二）：資訊系統",
    "備註": "",
    "全/半": "半",
    "全英語": "",
    "地點時間": "五 10-B 公館 教室自排",
    "學分": "3.0",
    "年": "3",
    "必/選": "選",
    "教師": "蔣宗哲 王科植",
    "班": "",
    "系所": "資工系（學）",
    "組": "",
    "英文課程名稱": "Special Topics on Computer Science (II) : Information Systems",
    "選修人數": "35",
    "開課代碼": "CSU0040",
    "開課序號": "2894",
    "限修人數": "50",
    "限修條件": "",
    "限性別": "",
  },
  {
    "MOOCS": "",
    "中文課程名稱": "影像處理",
    "備註": "",
    "全/半": "半",
    "全英語": "",
    "地點時間": "四 2-4 公館 理圖003",
    "學分": "3.0",
    "年": "3",
    "必/選": "選",
    "教師": "陳世旺",
    "班": "",
    "系所": "資工系（學）",
    "組": "",
    "英文課程名稱": "Image Processing",
    "選修人數": "65",
    "開課代碼": "CSU0041",
    "開課序號": "2895",
    "限修人數": "50",
    "限修條件": "",
    "限性別": "",
  },
  {
    "MOOCS": "",
    "中文課程名稱": "人工智慧",
    "備註": "",
    "全/半": "半",
    "全英語": "",
    "地點時間": "一 7 公館 Ｂ101, 四 8-9 公館 Ｂ101",
    "學分": "3.0",
    "年": "3",
    "必/選": "選",
    "教師": "林順喜",
    "班": "",
    "系所": "資工系（學）",
    "組": "",
    "英文課程名稱": "Artificial Intelligence",
    "選修人數": "63",
    "開課代碼": "CSU0042",
    "開課序號": "2896",
    "限修人數": "50",
    "限修條件": "",
    "限性別": "",
  },
]
`;
